MDEV-37072: Implement IS JSON predicate

KhaledR57 opened the pull request:
- [x] *The Jira issue number for this PR is: MDEV-37072*

<!--
An amazing description should answer some questions like:
1. What problem is the patch trying to solve?
2. If some output changed that is not visible in a test case, what was it looking like before the change and how it's looking with this patch applied?
3. Do you think this patch might introduce side-effects in other parts of the server?
-->
## Description

Add support for the SQL standard IS JSON predicate with the syntax:
```
<expr> IS [ NOT ] JSON
             [ { VALUE | ARRAY | OBJECT | SCALAR } ]
             [ { WITH  | WITHOUT } UNIQUE [ KEYS ] ]
```

The predicate allows checking if an expression is valid JSON and optionally constrains the JSON type (VALUE, ARRAY, OBJECT, SCALAR) and whether object keys are unique.

The implementation includes:
- Basic IS JSON validation
- Support for NOT operator
- Type constraints (VALUE, ARRAY, OBJECT, SCALAR)
- Unique keys constraint (WITH/WITHOUT UNIQUE KEYS)

## How can this PR be tested?

```bash
$ ./mtr main.func_json
```

## Basing the PR against the correct MariaDB version
- [X] *This is a new feature or a refactoring, and the PR is based against the `main` branch.*
- [ ] *This is a bug fix, and the PR is based against the earliest maintained branch in which the bug can be reproduced.*


## PR quality check
- [X] I checked the [CODING_STANDARDS.md](https://github.com/MariaDB/server/blob/-/CODING_STANDARDS.md) file and my PR conforms to this where appropriate.
- [X] For any trivial modifications to the PR, I am ok with the reviewer making the changes themselves.

andrelkin commented:
As to [the failing mdev-35046.test ](https://buildbot.mariadb.org/#/grid?branch=refs%2Fpull%2F4332%2Fhead) it be of great help to restore `51054ba173` headed branch for investigation. Sadly or remarkably the failure with 
```
Warning: Internal memory accounting error of 32960 bytes
```  happens at building only this branch.

Perhaps that could be done dear @KhaledR57  with help of `git reflog` etc?

KhaledR57 commented:
@andrelkin Thank you for the comment.
I've reverted to commit `51054ba173` and found the cause of the memory accounting error was a memory leak in `check_unique_keys` function. I was allocating keys with `my_malloc()` without providing `my_free` function during `my_hash_init()`.
The fix is to pass `my_free` to the hash initialization, so the hash table frees keys automatically in `my_hash_free()`.

The reason it doesn't occur in the last commit `f8ca79667` is that I had switched to `malloc()` instead of `my_malloc()` (I mistakenly thought `my_malloc` was the issue, not the missing `my_free`), which hid the error but didn’t actually fix the leak.

Regarding mdev-35046.test, I could not reproduce the specific failure, it passes locally.

mariadb-RuchaDeodhar commented on sql/item_jsonfunc.cc:
```diff
@@ -6262,3 +6262,142 @@ bool Item_func_json_object_to_array::fix_length_and_dec(THD *thd)
+    break;
+  }
+
+  while (json_scan_next(&je) == 0) {}
```
I suggest, instead of checking here whether the json is correctly formed ( has no error), check it inside check_unique_keys(). You are already iterating through the json in check_unique_keys(). Before returning anything, iterate through remaining json if left and adjust the return value accordingly (whether remaining json has error or not).
This way we avoid iterating through the json twice.

KhaledR57 replied:
I’ve added a check so that when `with_unique_keys` is true we just call `check_unique_keys()` (which already scans through the JSON and checks validity and uniqueness), so the unique-keys path avoids a second full scan.

mariadb-RuchaDeodhar commented on mysql-test/main/func_json.result:
```diff
@@ -5263,4 +5263,205 @@ SET @obj1='{ "a": 1,"b": 2,"c": 3}';
+{"a":1, "b":2, "c":3}	1	1	0	0
+{"a":1, "a":2, "a":3}	0	1	0	0
+[{"a":"1"},{"c":"2","c":"3"}]	0	0	0	1
+[{"a":"1"},{"b":"2","c":"3"}]	0	0	1	1
```
please check the results again. :)

mariadb-RuchaDeodhar commented on mysql-test/main/func_json.result:
```diff
@@ -5263,4 +5263,205 @@ SET @obj1='{ "a": 1,"b": 2,"c": 3}';
+json_col IS JSON OBJECT WITHOUT UNIQUE KEYS as object_without_unique
+FROM test_json;
+json_col	is_json	is_value	is_scalar	is_object	is_array	object_with_unique	object_without_unique
+{"name":"Alice", "age":25}	1	1	0	1	0	1	1
```
Result looks incorrect. Please fix. object_with_unique and object_without_unique cant be both true at the same time.
